### PR TITLE
fix-issue-ocp-22606

### PR DESCRIPTION
**Pull request by sandeepknd**
Hi @geliu2016 ,
Could you please have a look.
Please find the execution [log](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/4290/console)

### CODE REVIEW

File: features/upgrade/etcd/upgrade.feature [hunk 1]
```diff
@@ -13,26 +13,26 @@ Feature: basic verification for upgrade testing
...
-    #bugzilla - id=1979550
-    #Then status becomes :running of exactly 3 pods labeled:
-    #  | etcd_cluster=example |
-    When I run the :get admin command with:
-      | resource | EtcdCluster |
-    Then the step should succeed
-    And the output should match:
-      | example.* |
+    When I use the "openshift-etcd" project
+    And status becomes :running of 3 pods labeled:
```
Comment (geliu2016) [inline features/upgrade/etcd/upgrade.feature:17]: it will be better if check master node number, then check etcd member pods, because hard code 3 pods will fail in SNO test.

File: features/upgrade/etcd/upgrade.feature [hunk 2]
```diff
@@ -13,26 +13,26 @@ Feature: basic verification for upgrade testing
...
+      | is healthy: successfully committed proposal |
+    #login in etcd member pod, and add data in the form of key - value
+    When I run the :rsh client command with:
+      | pod         | <%= cb.etcdpod %> |
+      | command     | etcdctl           |
+      | command_arg | put               |
+      | command_arg | data-ocp26206     |
+      | command_arg | value-of-OCP26206 |
+    Then the output should contain:
+      | OK |
```
Comment (geliu2016) [inline features/upgrade/etcd/upgrade.feature:35]: it will be better add etcd-operator pods checking in ns: openshift-etcd-operator, and also etcd cluster operator checking(oc get co|grep etcd, check the status: True False False )

File: features/upgrade/etcd/upgrade.feature [hunk 3]
```diff
@@ -13,26 +13,26 @@ Feature: basic verification for upgrade testing
...
+      | command_arg | health            |
+    Then the output should contain 3 times:
+      | is healthy: successfully committed proposal |
+    #login in etcd member pod, and add data in the form of key - value
+    When I run the :rsh client command with:
+      | pod         | <%= cb.etcdpod %> |
+      | command     | etcdctl           |
+      | command_arg | put               |
+      | command_arg | data-ocp26206     |
+      | command_arg | value-of-OCP26206 |
```
Comment (geliu2016) [inline features/upgrade/etcd/upgrade.feature:33]: it will be better to execute get value after put the key: value, let check the key:value saved in etcd successfully before upgrade start.

**Comment by sandeepknd** [comment]
Hi @geliu2016 ,
Could you please have a look.
Please find the upgrade prepare execution [logs](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/4319/)

Please find the upgrade test execution logs : [logs](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/4321/console)

File: features/upgrade/etcd/upgrade.feature (hunk 3 above)
Comment (sandeepknd) [reply features/upgrade/etcd/upgrade.feature:38]: dine

File: features/upgrade/etcd/upgrade.feature [hunk 4]
```diff
@@ -13,26 +12,47 @@ Feature: basic verification for upgrade testing
...
-      | example.* |
+    Given I store the masters in the :masters clipboard
+    When I use the "openshift-etcd" project
+    Given I run the steps <%= cb.masters.length %> times:
+    """
+    And status becomes :running of 1 pods labeled:
+      | app=etcd |
+    And status becomes :running of 1 pods labeled:
+      | name=etcd-quorum-guard |
+    """
```
Comment (geliu2016) [inline features/upgrade/etcd/upgrade.feature:23]: @sandeepknd the logic is wrong, if cb.masters.length == 3, and if pod number == 1, this step still pass succ, because get same pod 3 times

**Comment by sandeepknd** [comment]
Hi @geliu2016 ,
Please find the successful execution [log](https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/4347/console)
Could you please have another look. Thanks for suggestions.

File: features/step_definitions/upgrade/etcd.rb [hunk 5]
```diff
@@ -75,4 +75,78 @@
   logger.info("### etcdCluster: #{name} is removed successfully from #{proj_name} namespace")
 end
 
+Then /^(the|all)? etcd pods outputs?( by order)? should( not)? (contain|match)? #{NUMBER} times:$/ do |all, in_order, negative, match_type, num, table|
```
Comment (pruan-rht) [inline features/step_definitions/upgrade/etcd.rb:78]: @geliu2016 @sandeepknd this step is unclear and perhaps over complicated.   For example what does the following mean?  ```And the etcd pods output should contain <%= cb.masters.length %> times```  What is the pattern it's watching over?

Comment (geliu2016) [reply features/step_definitions/upgrade/etcd.rb:78]: @pruan-rht this step will check etcd member/guard pod, which in "Running" status, if there are 3 member pods, so I want to get 3 Running, if there is 1 member pods in SNO, then I want to get 1 Running

Comment (pruan-rht) [reply features/step_definitions/upgrade/etcd.rb:78]: @geliu2016 ok...I would rephrase the step to make the purpose more clear.  Also, I would simplify the things to check to lessen the logic.   For example, the  (by order, and contain|match) added more logic to the step that don't really bring much more values.

Comment (geliu2016) [reply features/step_definitions/upgrade/etcd.rb:78]: @pruan-rht Great, thanks in advance

File: features/step_definitions/common.rb [hunk 6]
```diff
@@ -30,7 +30,7 @@
 # @param [Table] Array of the strings we are searching for
 # @note Checks whether the output contains/(does not contain) any of the
 # strings or regular expressions listed in the given table
-Then /^(the|all)? outputs?( by order)? should( not)? (contain|match)(?: (\d+) times)?:$/ do |all, in_order, negative, match_type, times, table|
+Then /^(the|all)? outputs?( by order)? should( not)? (contain|match)(?: #{NUMBER} times)?:$/ do |all, in_order, negative, match_type, times, table|
```
[the remark this answers is not in the document]
Comment (sandeepknd) [reply features/step_definitions/common.rb:33]: @pruan-rht ,
Actually I did a grep for the above function in both "verification tests" and  "cucushift" repo and the below combinations 
1. the outputs should contain "some string"
2. the outputs should match:  
    | some string   |
    
I tested these with the above changes and it worked fine.

[the remark this answers is not in the document]
Comment (sandeepknd) [reply features/step_definitions/common.rb]: Hi @pruan-rht ,
I have reverted the above change in the common.rb file.

File: features/step_definitions/upgrade/etcd.rb (hunk 5 above)
Comment (pruan-rht) [reply features/step_definitions/upgrade/etcd.rb:82]: @geliu2016 @sandeepknd after reviewing your testcase, you can simplifies by a lot by using existing methods.  I'll comment more inline... This new step is not needed.

File: features/upgrade/etcd/upgrade.feature [hunk 7]
```diff
@@ -13,26 +12,51 @@ Feature: basic verification for upgrade testing
...
-    Given I obtain test data file "admin/etcd-cluster.yaml"
-    When I run the :create client command with:
-      | f | etcd-cluster.yaml |
-    Then the step should succeed
-    #bugzilla - id=1979550
-    #Then status becomes :running of exactly 3 pods labeled:
-    #  | etcd_cluster=example |
+    Given I store the masters in the :masters clipboard
+    When I use the "openshift-etcd" project
+    When I run the :get admin command with:
```
Comment (pruan-rht) [inline features/upgrade/etcd/upgrade.feature:17]: @geliu2016 @sandeepknd you can replace the :get command (line 17-22) with the following
```
    And I store in the :pods clipboard the pods labeled:
      | app=etcd |
    Then the expression should be true> cb.pods.select {|p| p.ready_state_reachable?}.count == cb.masters.count
```

File: features/upgrade/etcd/upgrade.feature [hunk 8]
```diff
@@ -13,26 +12,51 @@ Feature: basic verification for upgrade testing
...
+      | resource | pods                              |
+      | l        | app=etcd                          |
+      | o        | jsonpath={.items[*].status.phase} |
+    And the etcd pods output should contain <%= cb.masters.length%> times:
+      | Running |
     When I run the :get admin command with:
```
Comment (pruan-rht) [inline features/upgrade/etcd/upgrade.feature:23]: @geliu2016 @sandeepknd you can replace line 23-28 with
```
And evaluation of `cb.pods[0].name` is stored in the :etcdpod clipboard
```

File: features/upgrade/etcd/upgrade.feature [hunk 9]
```diff
@@ -45,11 +69,44 @@ Feature: basic verification for upgrade testing
...
   @arm64 @amd64
-  Scenario: etcd-operator and cluster works well after upgrade
+    Scenario: etcd-operator and cluster works well after upgrade
     Given I switch to cluster admin pseudo user
-    When I use the "openshift-operators" project
-    Then status becomes :running of exactly 1 pods labeled:
-      | name=etcd-operator-alm-owned |
+    Given I store the masters in the :masters clipboard
+    When I use the "openshift-etcd" project
+    When I run the :get admin command with:
```
Comment (pruan-rht) [inline features/upgrade/etcd/upgrade.feature:76]: @geliu2016 @sandeepknd same replacement as I indicated earlier.
